File: codeforces/NumberTheory/Practice/Divisor.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Divisor {
    public static class FastReader {        // fast reader class
        public StringTokenizer tokenizer;
        public BufferedReader buffer;

        public FastReader() {this.buffer = new BufferedReader(new InputStreamReader(System.in));}       // Input reader

        public String next() {
            while(tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {tokenizer = new StringTokenizer(buffer.readLine());}
                catch(IOException e) {e.printStackTrace();}
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {return Integer.parseInt(next());}     // reading int
        public long nextLong() {return Long.parseLong(next());}     // reading long
    }

    public static void main(String[] args) {
        FastReader fast = new FastReader();
        final int n = fast.nextInt();
        int nums[] = new int[n];
        int firstNum = -1;
        for(int i = 0; i < n; i++) {
            nums[i] = fast.nextInt();
            // IMP- Since the divisors (1 and n) are also there, the max element is either x or y
            firstNum = Math.max(firstNum, nums[i]);
        }
        final StringBuilder builder = new StringBuilder();
        // We then find the second element
        builder.append(findSecondNumber(n, nums, firstNum)).append(" "+firstNum);
        System.out.print(builder);
    }

    public static int findSecondNumber(int n, int nums[], int firstNum) {
        Set<Integer> firstDivisors = new HashSet<>();
        for(int i = 0; i < n; i++)
            // Marking only one instance of the divisors
            if(firstNum % nums[i] == 0 && firstDivisors.add(nums[i]))   // Element not present, then add it
                nums[i] = -1;
        // IMP- Finding the max unmarked element as the second element
        return Arrays.stream(nums).max().orElse(Integer.MIN_VALUE);
    }
}

File: codeforces/NumberTheory/Practice/Divisor.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <climits>
using namespace std;

#define fast ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

int findSecondNumber(int n, vector<int>& nums, int firstNum) {
    unordered_set<int> firstDivisors;
    for(int i = 0; i < n; i++) {
        // IMP- If nums[i] divides firstNum and hasn't already been added (element not present)
        if(firstNum % nums[i] == 0 && firstDivisors.find(nums[i]) == firstDivisors.end()) {
            firstDivisors.insert(nums[i]);
            nums[i] = -1; // mark it
        }
    }
    // IMP- Find the maximum among unmarked values
    int secondNum = INT_MIN;
    for(int i = 0; i < n; i++) {
        if(nums[i] != -1)
            secondNum = max(secondNum, nums[i]);
    }
    return secondNum;
}

int main() {
    fast;
    int n;
    cin >> n;
    vector<int> nums(n);
    int firstNum = -1;
    for(int i = 0; i < n; i++) {
        cin >> nums[i];
        firstNum = max(firstNum, nums[i]);
    }
    int secondNum = findSecondNumber(n, nums, firstNum);
    cout << secondNum << " " << firstNum << "\n";

    return 0;
}
